File: app/src/main/java/com/huynd/skyobserver/presenters/cheapestflight/FlightWithCheapestPriceResultPresenterImpl.java
package com.huynd.skyobserver.presenters.cheapestflight;

import com.huynd.skyobserver.models.cheapestflight.CountryPriceInfo;
import com.huynd.skyobserver.models.cheapestflight.FlightWithCheapestPriceResultModel;
import com.huynd.skyobserver.services.PricesAPI;
import com.huynd.skyobserver.views.cheapestflight.FlightWithCheapestPriceResultView;

import java.util.List;

/**
 * Created by HuyND on 4/21/2018.
 */

public class FlightWithCheapestPriceResultPresenterImpl implements
        FlightWithCheapestPriceResultPresenter,
        FlightWithCheapestPriceResultModel.EventListener {

    private FlightWithCheapestPriceResultView mView;
    private FlightWithCheapestPriceResultModel mModel;
    private PricesAPI mPricesAPI;

    public FlightWithCheapestPriceResultPresenterImpl(FlightWithCheapestPriceResultView view,
                                                      PricesAPI pricesAPI) {
        mView = view;
        mPricesAPI = pricesAPI;

        mModel = new FlightWithCheapestPriceResultModel();
        mModel.setEventListener(this);
    }

    @Override
    public void notifyInvalidDate() {
        // TODO
    }

    @Override
    public void onGetPricesResponse(List<CountryPriceInfo> countryPriceInfos) {
        // TODO
    }
}

File: app/src/main/java/com/huynd/skyobserver/presenters/cheapestflight/FlightWithCheapestPriceResultPresenterImpl.kt
package com.huynd.skyobserver.presenters.cheapestflight

import com.huynd.skyobserver.models.cheapestflight.CountryPriceInfo
import com.huynd.skyobserver.models.cheapestflight.FlightWithCheapestPriceModel
import com.huynd.skyobserver.services.PricesAPI
import com.huynd.skyobserver.views.cheapestflight.FlightWithCheapestPriceResultView

/**
 * Created by HuyND on 4/21/2018.
 */

class FlightWithCheapestPriceResultPresenterImpl(
        private val mView: FlightWithCheapestPriceResultView,
        private val mPricesAPI: PricesAPI) : FlightWithCheapestPriceResultPresenter,
        FlightWithCheapestPriceModel.EventListener {
    private val mModel = FlightWithCheapestPriceModel()

    init {
        mModel.setEventListener(this)
    }

    override fun getPrices(yearOutbound: Int, monthOutbound: Int, dayOutbound: Int,
                           yearInbound: Int, monthInbound: Int, dayInbound: Int,
                           port: String, isReturnTrip: Boolean) {
        mView.showLoadingDialog()
        mModel.getPrices(mPricesAPI,
                yearOutbound, monthOutbound, dayOutbound,
                yearInbound, monthInbound, dayInbound,
                port, isReturnTrip)
    }

    override fun notifyInvalidDate() {
        mView.showInvalidDateDialog()
    }

    override fun onGetPricesResponse(listCountryPriceInfo: List<CountryPriceInfo>) {
        mView.run {
            updateListViewInboundPrices(listCountryPriceInfo)
            dismissLoadingDialog()
        }
    }
}
